File: src/main/java/database/UserDatabase.java
package database;

import com.koenig.commonModel.Family;
import com.koenig.commonModel.Item;
import com.koenig.commonModel.ItemType;
import com.koenig.commonModel.User;
import com.koenig.commonModel.database.DatabaseItem;
import com.koenig.commonModel.database.UserService;

import java.sql.Connection;
import java.sql.SQLException;
import java.util.List;

public class UserDatabase extends Database {
    private UserTable userTable;
    private FamilyTable familyTable;

    public UserDatabase(Connection connection) {
        super(connection);
    }

    @Override
    protected ItemTable getItemTable(Item item) throws SQLException {
        switch (ItemType.Companion.fromItem(item)) {
            case FAMILY:
                return familyTable;
            case USER:
                return userTable;
            default:
                throw new SQLException("Unsupported item in user database");
        }
    }

    public void start() throws SQLException {
        userTable = new UserTable(getConnection());
        familyTable = new FamilyTable(getConnection(), getUserService());
        getTables().add(userTable);
        getTables().add(familyTable);
        createAllTables();

        // TEST CODE

    }

    public UserService getUserService() {
        return id -> userTable.getFromId(id);
    }

    public void stop() throws SQLException {
        if (getConnection() != null)
            getConnection().close();
    }

    public void addUser(User user, String id) throws SQLException {
        if (user.getName().isEmpty()) {
            throw new SQLException("Username is empty");
        }

        DatabaseItem<User> databaseItem = new DatabaseItem<>(user, id);
        userTable.add(databaseItem);
    }

    public List<User> getAllUser() {
        return userTable.toItemList(userTable.getAll());
    }

    public void addFamily(Family family, String id) throws SQLException {
        DatabaseItem<Family> databaseItem = new DatabaseItem<>(family, id);
        familyTable.add(databaseItem);
    }

    public List<Family> getAllFamilys() {
        return familyTable.toItemList(familyTable.getAll());
    }

    public void addUserToFamily(String familyName, String userId) throws SQLException {

        Family family = familyTable.getFamilyByName(familyName);
        if (family == null) {
            throw new SQLException("Family does not exist");
        }

        startTransaction(new Transaction() {
            @Override
            public void run() throws SQLException {
                familyTable.addUserToFamily(family, userId);
                userTable.addFamileToUser(family.getName(), userId);
            }
        });

    }




    public Family getFamilyByName(String familyName) throws SQLException {
        return familyTable.getFamilyByName(familyName);
    }

    public User getUserById(String userId) throws SQLException {

        return userTable.getFromId(userId);
    }

    public UserTable getUserTable() {
        return userTable;
    }

    public String getFamilyIdFromUser(String userId) throws SQLException {
        Family familyFromUser = getFamilyFromUser(userId);
        if (familyFromUser == null) return null;
        return familyFromUser.getId();
    }

    public List<User> getFamilyMemberFrom(String userId) throws SQLException {
        Family family = getFamilyFromUser(userId);
        return family.getUsers();
    }

    private Family getFamilyFromUser(String userId) throws SQLException {
        User user = getUserById(userId);
        if (user == null) {
            throw new SQLException("User not found with id " + userId);
        }

        String family = user.getFamily();
        if (family == null) {
            throw new SQLException("Family not found from " + userId);
        }

        return familyTable.getFamilyByName(family);
    }
}

File: src/main/java/database/UserDatabase.kt
package database

import com.koenig.commonModel.Family
import com.koenig.commonModel.Item
import com.koenig.commonModel.ItemType
import com.koenig.commonModel.User
import com.koenig.commonModel.database.DatabaseItem
import java.sql.Connection
import java.sql.SQLException

class UserDatabase(connection: Connection) : Database(connection) {
    var userTable: UserTable = UserTable(connection)
    var familyTable: FamilyTable = FamilyTable(connection, userService)

    val userService: (String) -> User?
        get() = { id -> userTable.getFromId(id) }

    val allUser: List<User>
        get() = userTable.toItemList(userTable.all)

    val allFamilys: List<Family>
        get() = familyTable.toItemList(familyTable.all)

    @Suppress("UNCHECKED_CAST")
    @Throws(SQLException::class)
    override fun getItemTable(item: Item): ItemTable<Item> {
        return when (ItemType.fromItem(item)) {
            ItemType.FAMILY -> familyTable as ItemTable<Item>
            ItemType.USER -> userTable as ItemTable<Item>
            else -> throw SQLException("Unsupported item in user database")
        }
    }

    @Throws(SQLException::class)
    fun start() {
        tables.add(userTable)
        tables.add(familyTable)
        createAllTables()
    }

    @Throws(SQLException::class)
    fun stop() {
        connection.close()
    }

    @Throws(SQLException::class)
    fun addUser(user: User, id: String) {
        if (user.name.isEmpty()) {
            throw SQLException("Username is empty")
        }

        val databaseItem = DatabaseItem(user, id)
        userTable.add(databaseItem)
    }

    @Throws(SQLException::class)
    fun addFamily(family: Family, id: String) {
        val databaseItem = DatabaseItem(family, id)
        familyTable.add(databaseItem)
    }

    @Throws(SQLException::class)
    fun addUserToFamily(familyName: String, userId: String) {

        val family = familyTable.getFamilyByName(familyName) ?: throw SQLException("Family does not exist")

        startTransaction({
            familyTable.addUserToFamily(family, userId)
            userTable.addFamileToUser(family.name, userId)
        })

    }


    @Throws(SQLException::class)
    fun getFamilyByName(familyName: String): Family? {
        return familyTable.getFamilyByName(familyName)
    }

    @Throws(SQLException::class)
    fun getUserById(userId: String): User? {

        return userTable.getFromId(userId)
    }

    @Throws(SQLException::class)
    fun getFamilyIdFromUser(userId: String): String {
        val familyFromUser = getFamilyFromUser(userId) ?: throw SQLException("Family from user not found")
        return familyFromUser.id
    }

    @Throws(SQLException::class)
    fun getFamilyMemberFrom(userId: String): List<User> {
        val family = getFamilyFromUser(userId)
        return family!!.users
    }

    @Throws(SQLException::class)
    fun getFamilyFromUser(userId: String): Family? {
        val user = getUserById(userId) ?: throw SQLException("User not found with id $userId")

        val family = user.family

        return familyTable.getFamilyByName(family)
    }
}
